### PR TITLE
Adding CSV header support to standard library. All changes are non-br…

### PR DESCRIPTION
**Pull request by hamin**
…eaking to current CSV api. References #1016

### CODE REVIEW

File: spec/std/csv/csv_parse_spec.cr [hunk 1]
```diff
@@ -22,6 +22,13 @@ describe CSV do
         ])
     end
 
+    it "parses two rows with a header" do
+      CSV.parse("name,age\nharis,21\nbobby,35\n", header_row = true).should eq([
```
Comment (will) [inline spec/std/csv/csv_parse_spec.cr]: This should be `header_row: true`

File: src/csv/csv.cr [hunk 2]
```diff
@@ -44,6 +44,26 @@ module CSV
...
+  # ```
+  #
+  # Output:
+  #
+  # ```
+  # {"name" => "haris", "age" => "21"}
+  # {"name" => "bobby", "age" => "35"}
+  # ```
+  def self.each_row_with_header(string_or_io : String | IO)
+    Parser.new(string_or_io, header_row = true).each_row_with_header do |row|
```
Comment (will) [inline src/csv/csv.cr]: and here

File: src/csv/csv.cr [hunk 3]
```diff
@@ -55,6 +75,17 @@ module CSV
...
 
+  # Returns an `Iterator` of `Hash(String,String)` over a CSV's rows.
+  #
+  # ```
+  # rows = CSV.each_row_with_header("name,age\nharis,21\nbobby,35")
+  # rows.next #=> {"name" => "haris", "age" => "21"}
+  # rows.next #=> {"name" => "bobby", "age" => "35"}
+  # ```
+  def self.each_row_with_header(string_or_io : String | IO)
+    Parser.new(string_or_io, header_row = true).each_row_with_header
```
Comment (will) [inline src/csv/csv.cr]: and here

**Comment by hamin** [comment]
@will done, wasn't sure which was the more appropriate style :)

**Comment by will** [comment]
It's not a style thing, it's a bug to use `=` at call site, because what's happening is you're defining a new local variable and assiging it, and then that's being passed into the the condition.

```
/tmp ➤ cat call.cr
def foo(one=1, two=2, three=3)
  {one, two, three}
end

pp foo()
pp foo(one: 6, three: 1)
pp foo(one = 6, three = 1)


/tmp ➤ crystal call.cr
foo = {1, 2, 3}
foo(one: 6, three: 1) = {6, 2, 1}
foo(one = 6, three = 1) = {6, 1, 3}
```

In this case the second position and the name were the same so it still worked.

**Comment by asterite** [comment]
Thanks!

Some notes:
1. Changing a method to suddenly returning a union breaks all code in existence. Additionally, dealing with unions is harder than dealing with single types. Also, if you want to parse with headers you know you'll deal with a hash, otherwise with an array. So the best thing to do here is to have two different methods, maybe `parse_with_headers`, or something like that.
2. There's absolutely no need to change the Parser for this.
3. Creating a hash for each row is probably overkill, but can work. What I'd really like is some wrapper that has a reference to the first row (headers, and indices) and then a sort of CSV::Row that has a reference to that wrapper that lets you get rows by name or index. That way there's no need to create intermediate hashes. I say this because I had to deal with large CSVs and it would be nice to be able to deal with this without allocating needless memory.

**Comment by hamin** [comment]
@will ah I see. This is really my first foray in Crystal so still learning :)

@asterite Thanks for your feedback. Again this is my first foray so apologies for my ignorance on best practices. Regarding your 3 points:

1. IMHO me CSV.parse should really be able to handle any RFC compliant CSV including headers. That's why I thought it made sense to keep that method the same "interface" (pardon the poor terminology here) with the optional argument of whether or not the CSV file contains a header or not. This is the approach the current Ruby CSV std library takes (which is quite well written, JEG II the author is awesome :) ). But I do see what you're saying about returning a union. Correct me if I'm wrong (I'm still new to Crystal here) but are you referring to this: Array(Array(String)) | Array(Hash(String,String)) ? I think that's what you're referring to and I see where you're coming from. However I'm not sure "breaks all code in existence" :) . I made special care to make sure none of the existing tests fail and I don't see where this could be a problem. Again I most definitely might be missing something here. Please do let me know.

2. I guess again it comes down to I was trying to keep the end user API clean. The only way I can think of not changing the Parser for this is to perhaps right another parser that handles the headers. My Parser additions mostly added new methods and didn't change the existing methods that much. What do you suggest might be best here?

3. Completely agree with you on there being a CSV:Row object of some sort. That's actually how the Ruby CSV parser is kinda organized. But I think at that point this would be an major rewrite. I was trying to add support for headers without having to do that at the moment per se. Do you have any specific suggestions of stuff I can be doing better here? 

Again thanks for your patience...still learning about crystal :)

**Comment by hamin** [comment]
@asterite Here's the ruby implementation https://github.com/ruby/ruby/blob/trunk/lib/csv.rb Do you think its worth attempting a rewrite of the Crystal CSV lib making it closer to the ruby implementation. As a user faced api I would love to have something quite similar to the Ruby version. Not sure how big of an undertaking that would be at the moment

**Comment by asterite** [comment]
@hamin I finally decided to do it myself, with some advice from @waj. We want to ensure an API that allows for optimal performance when parsing a CSV, and the API we came up with allows just that, only needing one array instance in memory for parsing a whole CSV file.

**Comment by asterite** [comment]
@hamin But thanks for bringing this up! Your pull request and #1016 pushed at to solving this issue.

**Comment by hamin** [comment]
@asterite sure no problem :). I would have done it the same way I think you were suggesting but as I said I think it would be a rewrite of the API which I think we both agree is best. Do check out the ruby version which is actually very efficient and quite well written. Glad I could help in some way :). Will keep my eyes peeled for any more ways i can help out :)

**Pull request closed without merging**